webservices/server-integration/src/main/java/org/jboss/as/webservices/publish/EndpointPublisherImpl.java

```diff
@@ -175,22 +178,25 @@ protected void doDeploy(ServiceTarget target, DeploymentUnit unit) {
      * @throws Exception
      */
     protected Context doPublish(ServiceTarget target, DeploymentUnit unit) throws Exception {
-        Deployment deployment = unit.getAttachment(WSAttachmentKeys.DEPLOYMENT_KEY);
-        List<Endpoint> endpoints = deployment.getService().getEndpoints();
+        final Deployment deployment = unit.getAttachment(WSAttachmentKeys.DEPLOYMENT_KEY);
+        final List<Endpoint> endpoints = deployment.getService().getEndpoints();
         //If we're running in a Service, that will already have proper dependencies set on the installed endpoint services,
         //otherwise we need to explicitly wait for the endpoint services to be started before creating the webapp.
         if (!runningInService) {
             final ServiceRegistry registry = unit.getServiceRegistry();
-            final StabilityMonitor monitor = new StabilityMonitor();
+            final UninterruptibleCountDownLatch latch = new UninterruptibleCountDownLatch(endpoints.size());
+            final LifecycleListener listener = new LifecycleListener() {
+                @Override
+                public void handleEvent(final ServiceController<?> controller, final LifecycleEvent event) {
+                    if (event == LifecycleEvent.UP) latch.countDown();
+                }
+            };
+            ServiceName serviceName;
             for (Endpoint ep : endpoints) {
-                final ServiceName serviceName = EndpointService.getServiceName(unit, ep.getShortName());
-                monitor.addController(registry.getRequiredService(serviceName));
-            }
-            try {
-                monitor.awaitStability();
-            } finally {
-                monitor.clear();
+                serviceName = EndpointService.getServiceName(unit, ep.getShortName());
+                registry.getRequiredService(serviceName).addListener(listener);
```

Comment:
Shouldn't this lifecycle listener be removed in a finally block?

Reply:
Of course it must be removed @pferraro - thanks for spotting that out.